dgraphtest/image.go

```diff
@@ -16,6 +16,155 @@
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/pkg/errors"
+)
+
+var (
+	homePath = os.Getenv("HOME") + "/dgraph"
+	repoDir  = homePath + "/binaries"
+)
+
+const (
+	relativeDir   = "binaries"
+	dgraphRepoUrl = "https://github.com/dgraph-io/dgraph.git"
+)
+
 func (c *LocalCluster) dgraphImage() string {
 	return "dgraph/dgraph:local"
 }
+
+func (c *LocalCluster) setupBinary(version string, logger Logger) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(time.Minute*10))
+	defer cancel()
+	absPath, err := filepath.Abs(repoDir)
+	if err != nil {
+		return errors.Wrap(err, "error while getting absolute path")
+	}
+	repo, err := git.PlainOpen(absPath)
+	if err != nil && err == git.ErrRepositoryNotExists {
+		logger.Logf("cloning repo")
+		repo, err = git.PlainCloneContext(ctx, absPath, false, &git.CloneOptions{
+			URL: dgraphRepoUrl,
+		})
+		if err != nil {
+			return errors.Wrap(err, "error while checking out dgraph git repo")
+
+		}
+	} else if err != nil {
+		return errors.Wrap(err, "error while opening git repo")
+	} else {
+		if err := repo.Fetch(&git.FetchOptions{}); err != nil && err != git.NoErrAlreadyUpToDate {
+			return errors.Wrap(err, "error while fetching git repo")
+		}
+	}
+	var hash plumbing.Hash
+	if strings.Contains(version, "v") {
+		tagRef, err := repo.Tag(version)
+		if err != nil {
+			return errors.Wrap(err, "error while getting tagref of version")
+		}
+		hash = tagRef.Hash()
+	} else {
+		hash = plumbing.NewHash(version)
+	}
+
+	// checkout repo with specific tag
+	if err := checkoutGitRepo(repo, hash, "branch"+version); err != nil {
+		return err
+	}
+	absPath, err = filepath.Abs(relativeDir)
+	if err != nil {
+		return errors.Wrap(err, "error while getting absolute path")
+	}
+	// make dgraph of specific version
+	logger.Logf("making binary")
+	if err := makeDgraphBinary(repoDir, absPath, version); err != nil {
+		return err
+	}
+	err = copy(relativeDir+"/dgraph_"+version, c.tempDir+"/dgraph")
+	if err != nil {
+		return errors.Wrap(err, "error while copying dgraph binary into temp dir")
+	}
+	return nil
+}
+
+func checkoutGitRepo(repo *git.Repository, hash plumbing.Hash, branchName string) error {
+	worktree, err := repo.Worktree()
+	if err != nil {
+		return errors.Wrap(err, "error getting git repo work tree")
+
+	}
+	err = worktree.Checkout(&git.CheckoutOptions{
+		Hash: hash,
+	})
+	if err != nil && strings.Contains(err.Error(), "already exists") {
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "error getting while checking out git repo")
+	}
+	return nil
+}
+
+func makeDgraphBinary(dir, binaryDir, version string) error {
+	cmd := exec.Command("make", "dgraph")
+	cmd.Dir = dir + "/dgraph"
+	if err := cmd.Run(); err != nil {
+		return errors.Wrap(err, "error getting while building dgraph binary")
+	}
+	err := copy(dir+"/dgraph"+"/dgraph", binaryDir+"/dgraph_"+version)
```

Comment:
use sprintf for build binary names. Maybe create a small function for it so that we can change the naming easily in future if needed.